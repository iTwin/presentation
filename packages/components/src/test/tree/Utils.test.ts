/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

import { expect } from "chai";
import { LabelDefinition, Node } from "@itwin/presentation-common";
import { PageOptions } from "@itwin/components-react";
import {
  createPartialTreeNodeItem, createTreeNodeItem, createTreeNodeItems, pageOptionsUiToPresentation,
} from "../../presentation-components/tree/Utils";
import { createTestECClassGroupingNodeKey, createTestECInstancesNode } from "../_helpers/Hierarchy";

describe("Utils", () => {
  describe("createTreeNodeItem", () => {
    it("creates tree node", () => {
      const node = createTestECInstancesNode();
      const treeNode = createTreeNodeItem(node);
      expect(treeNode).to.matchSnapshot();
    });

    it("creates tree node with extended data", () => {
      const node = { ...createTestECInstancesNode(), extendedData: { test: "value" } };
      const treeNode = createTreeNodeItem(node);
      expect(treeNode.extendedData!.test).to.eq("value");
    });

    it("creates tree node with parent id", () => {
      const node = createTestECInstancesNode();
      const parentId = "test_parent_id";
      const treeNode = createTreeNodeItem(node, parentId);
      expect(treeNode).to.matchSnapshot();
    });

    it("creates tree node with custom label styles", () => {
      const node = createTestECInstancesNode();
      node.fontStyle = "Bold Italic"; // eslint-disable-line deprecation/deprecation
      const treeNode = createTreeNodeItem(node);
      expect(treeNode).to.matchSnapshot();
    });

    it("creates auto expanded tree node", () => {
      const node = createTestECInstancesNode({ isExpanded: true });
      const treeNode = createTreeNodeItem(node);
      expect(treeNode).to.matchSnapshot();
    });

    it("appends grouped nodes count if requested", () => {
      const node: Node = {
        key: createTestECClassGroupingNodeKey({ groupedInstancesCount: 999 }),
        label: LabelDefinition.fromLabelString("test"), // eslint-disable-line @itwin/no-internal
      };
      const treeNode = createTreeNodeItem(node, undefined, { appendChildrenCountForGroupingNodes: true });
      expect(treeNode).to.matchSnapshot();
    });

    it("uses provided callback to customize tree node", () => {
      const node = createTestECInstancesNode();
      const treeNode = createTreeNodeItem(node, undefined, {
        customizeTreeNodeItem: (item) => {
          item.icon = "custom-icon";
          item.description = "custom-description";
        },
      });
      expect(treeNode).to.matchSnapshot();
    });
  });

  describe("createPartialTreeNodeItem", () => {
    it("assigns item id and label from loaded node", () => {
      const node = createPartialTreeNodeItem(
        {
          key: { type: "", version: 0, pathFromRoot: [] },
          label: LabelDefinition.fromLabelString("test"), // eslint-disable-line @itwin/no-internal
        },
        undefined,
        {},
      );
      expect(node.id).not.to.be.undefined;
      expect(node.label).not.to.be.undefined;
      expect(node.key).not.to.be.undefined;
    });

    it("does not set a presentation tree node key when input does not have a key", () => {
      const node = createPartialTreeNodeItem({}, undefined, {});
      expect(node.key).to.be.undefined;
    });

    it("uses provided callback to customize tree node", () => {
      const treeNode = createPartialTreeNodeItem(
        {
          key: { type: "", version: 0, pathFromRoot: [] },
          label: LabelDefinition.fromLabelString("test"), // eslint-disable-line @itwin/no-internal
        }, undefined,
        {
          customizeTreeNodeItem: (item) => {
            item.icon = "custom-icon";
            item.description = "custom-description";
          },
        },
      );
      expect(treeNode).to.matchSnapshot();
    });
  });

  describe("createTreeNodeItems", () => {
    it("creates tree nodes", () => {
      const nodes = [createTestECInstancesNode(), createTestECInstancesNode()];
      const treeNode = createTreeNodeItems(nodes);
      expect(treeNode).to.matchSnapshot();
    });

    it("creates tree nodes with parentId", () => {
      const nodes = [createTestECInstancesNode(), createTestECInstancesNode()];
      const parentId = "test_parent_id";
      const treeNode = createTreeNodeItems(nodes, parentId);
      expect(treeNode).to.matchSnapshot();
    });

  });

  describe("pageOptionsUiToPresentation", () => {
    it("returns undefined if passed undefined parameter", () => {
      const result = pageOptionsUiToPresentation(undefined);
      expect(result).to.be.equal(undefined);
    });

    it("converts ui page options to presentation page options", () => {
      const size = 10;
      const start = 2;
      const pageOptions: PageOptions = { size, start };
      const result = pageOptionsUiToPresentation(pageOptions);

      expect(result).to.not.be.undefined;
      expect(result!.size).to.be.equal(size);
      expect(result!.start).to.be.equal(start);
    });
  });
});
