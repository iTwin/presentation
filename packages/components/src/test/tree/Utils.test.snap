// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils createPartialTreeNodeItem uses provided callback to customize tree node 1`] = `
Object {
  "description": "custom-description",
  "icon": "custom-icon",
  "id": "",
  "key": Object {
    "pathFromRoot": Array [],
    "type": "",
    "version": 0,
  },
  "label": PropertyRecord {
    "autoExpand": undefined,
    "description": undefined,
    "extendedData": undefined,
    "isDisabled": undefined,
    "isMerged": undefined,
    "isReadonly": undefined,
    "links": undefined,
    "property": Object {
      "displayLabel": "Label",
      "name": "node_label",
      "typename": "string",
    },
    "value": Object {
      "displayValue": "test",
      "value": "test",
      "valueFormat": 0,
    },
  },
}
`;

exports[`Utils createTreeNodeItem appends grouped nodes count if requested 1`] = `
Object {
  "id": "hash",
  "key": Object {
    "className": "TestSchema:GroupedClass",
    "groupedInstancesCount": 999,
    "pathFromRoot": Array [
      "hash",
    ],
    "type": "ECClassGroupingNode",
    "version": 2,
  },
  "label": PropertyRecord {
    "autoExpand": undefined,
    "description": undefined,
    "extendedData": undefined,
    "isDisabled": undefined,
    "isMerged": undefined,
    "isReadonly": undefined,
    "links": undefined,
    "property": Object {
      "displayLabel": "Label",
      "name": "node_label",
      "typename": "composite",
    },
    "value": Object {
      "displayValue": "test (999)",
      "value": Object {
        "parts": Array [
          Object {
            "displayValue": "test",
            "rawValue": "test",
            "typeName": "string",
          },
          Object {
            "displayValue": "(999)",
            "rawValue": "(999)",
            "typeName": "string",
          },
        ],
        "separator": " ",
      },
      "valueFormat": 0,
    },
  },
}
`;

exports[`Utils createTreeNodeItem creates auto expanded tree node 1`] = `
Object {
  "autoExpand": true,
  "description": "testDescription",
  "id": "childHash/parentHash",
  "key": Object {
    "instanceKeys": Array [
      Object {
        "className": "TestSchema:TestClass",
        "id": "0x1",
      },
      Object {
        "className": "TestSchema:TestClass",
        "id": "0x1",
      },
    ],
    "pathFromRoot": Array [
      "parentHash",
      "childHash",
    ],
    "type": "ECInstancesNode",
    "version": 2,
  },
  "label": PropertyRecord {
    "autoExpand": undefined,
    "description": undefined,
    "extendedData": undefined,
    "isDisabled": undefined,
    "isMerged": undefined,
    "isReadonly": undefined,
    "links": undefined,
    "property": Object {
      "displayLabel": "Label",
      "name": "node_label",
      "typename": "string",
    },
    "value": Object {
      "displayValue": "test label",
      "value": "test_label",
      "valueFormat": 0,
    },
  },
}
`;

exports[`Utils createTreeNodeItem creates tree node 1`] = `
Object {
  "description": "testDescription",
  "id": "childHash/parentHash",
  "key": Object {
    "instanceKeys": Array [
      Object {
        "className": "TestSchema:TestClass",
        "id": "0x1",
      },
      Object {
        "className": "TestSchema:TestClass",
        "id": "0x1",
      },
    ],
    "pathFromRoot": Array [
      "parentHash",
      "childHash",
    ],
    "type": "ECInstancesNode",
    "version": 2,
  },
  "label": PropertyRecord {
    "autoExpand": undefined,
    "description": undefined,
    "extendedData": undefined,
    "isDisabled": undefined,
    "isMerged": undefined,
    "isReadonly": undefined,
    "links": undefined,
    "property": Object {
      "displayLabel": "Label",
      "name": "node_label",
      "typename": "string",
    },
    "value": Object {
      "displayValue": "test label",
      "value": "test_label",
      "valueFormat": 0,
    },
  },
}
`;

exports[`Utils createTreeNodeItem creates tree node with custom label styles 1`] = `
Object {
  "description": "testDescription",
  "id": "childHash/parentHash",
  "key": Object {
    "instanceKeys": Array [
      Object {
        "className": "TestSchema:TestClass",
        "id": "0x1",
      },
      Object {
        "className": "TestSchema:TestClass",
        "id": "0x1",
      },
    ],
    "pathFromRoot": Array [
      "parentHash",
      "childHash",
    ],
    "type": "ECInstancesNode",
    "version": 2,
  },
  "label": PropertyRecord {
    "autoExpand": undefined,
    "description": undefined,
    "extendedData": undefined,
    "isDisabled": undefined,
    "isMerged": undefined,
    "isReadonly": undefined,
    "links": undefined,
    "property": Object {
      "displayLabel": "Label",
      "name": "node_label",
      "typename": "string",
    },
    "value": Object {
      "displayValue": "test label",
      "value": "test_label",
      "valueFormat": 0,
    },
  },
  "style": Object {
    "isBold": true,
    "isItalic": true,
  },
}
`;

exports[`Utils createTreeNodeItem creates tree node with parent id 1`] = `
Object {
  "description": "testDescription",
  "id": "childHash/parentHash",
  "key": Object {
    "instanceKeys": Array [
      Object {
        "className": "TestSchema:TestClass",
        "id": "0x1",
      },
      Object {
        "className": "TestSchema:TestClass",
        "id": "0x1",
      },
    ],
    "pathFromRoot": Array [
      "parentHash",
      "childHash",
    ],
    "type": "ECInstancesNode",
    "version": 2,
  },
  "label": PropertyRecord {
    "autoExpand": undefined,
    "description": undefined,
    "extendedData": undefined,
    "isDisabled": undefined,
    "isMerged": undefined,
    "isReadonly": undefined,
    "links": undefined,
    "property": Object {
      "displayLabel": "Label",
      "name": "node_label",
      "typename": "string",
    },
    "value": Object {
      "displayValue": "test label",
      "value": "test_label",
      "valueFormat": 0,
    },
  },
  "parentId": "test_parent_id",
}
`;

exports[`Utils createTreeNodeItem uses provided callback to customize tree node 1`] = `
Object {
  "description": "custom-description",
  "icon": "custom-icon",
  "id": "childHash/parentHash",
  "key": Object {
    "instanceKeys": Array [
      Object {
        "className": "TestSchema:TestClass",
        "id": "0x1",
      },
      Object {
        "className": "TestSchema:TestClass",
        "id": "0x1",
      },
    ],
    "pathFromRoot": Array [
      "parentHash",
      "childHash",
    ],
    "type": "ECInstancesNode",
    "version": 2,
  },
  "label": PropertyRecord {
    "autoExpand": undefined,
    "description": undefined,
    "extendedData": undefined,
    "isDisabled": undefined,
    "isMerged": undefined,
    "isReadonly": undefined,
    "links": undefined,
    "property": Object {
      "displayLabel": "Label",
      "name": "node_label",
      "typename": "string",
    },
    "value": Object {
      "displayValue": "test label",
      "value": "test_label",
      "valueFormat": 0,
    },
  },
}
`;

exports[`Utils createTreeNodeItems creates tree nodes 1`] = `
Array [
  Object {
    "description": "testDescription",
    "id": "childHash/parentHash",
    "key": Object {
      "instanceKeys": Array [
        Object {
          "className": "TestSchema:TestClass",
          "id": "0x1",
        },
        Object {
          "className": "TestSchema:TestClass",
          "id": "0x1",
        },
      ],
      "pathFromRoot": Array [
        "parentHash",
        "childHash",
      ],
      "type": "ECInstancesNode",
      "version": 2,
    },
    "label": PropertyRecord {
      "autoExpand": undefined,
      "description": undefined,
      "extendedData": undefined,
      "isDisabled": undefined,
      "isMerged": undefined,
      "isReadonly": undefined,
      "links": undefined,
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "test label",
        "value": "test_label",
        "valueFormat": 0,
      },
    },
  },
  Object {
    "description": "testDescription",
    "id": "childHash/parentHash",
    "key": Object {
      "instanceKeys": Array [
        Object {
          "className": "TestSchema:TestClass",
          "id": "0x1",
        },
        Object {
          "className": "TestSchema:TestClass",
          "id": "0x1",
        },
      ],
      "pathFromRoot": Array [
        "parentHash",
        "childHash",
      ],
      "type": "ECInstancesNode",
      "version": 2,
    },
    "label": PropertyRecord {
      "autoExpand": undefined,
      "description": undefined,
      "extendedData": undefined,
      "isDisabled": undefined,
      "isMerged": undefined,
      "isReadonly": undefined,
      "links": undefined,
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "test label",
        "value": "test_label",
        "valueFormat": 0,
      },
    },
  },
]
`;

exports[`Utils createTreeNodeItems creates tree nodes with parentId 1`] = `
Array [
  Object {
    "description": "testDescription",
    "id": "childHash/parentHash",
    "key": Object {
      "instanceKeys": Array [
        Object {
          "className": "TestSchema:TestClass",
          "id": "0x1",
        },
        Object {
          "className": "TestSchema:TestClass",
          "id": "0x1",
        },
      ],
      "pathFromRoot": Array [
        "parentHash",
        "childHash",
      ],
      "type": "ECInstancesNode",
      "version": 2,
    },
    "label": PropertyRecord {
      "autoExpand": undefined,
      "description": undefined,
      "extendedData": undefined,
      "isDisabled": undefined,
      "isMerged": undefined,
      "isReadonly": undefined,
      "links": undefined,
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "test label",
        "value": "test_label",
        "valueFormat": 0,
      },
    },
    "parentId": "test_parent_id",
  },
  Object {
    "description": "testDescription",
    "id": "childHash/parentHash",
    "key": Object {
      "instanceKeys": Array [
        Object {
          "className": "TestSchema:TestClass",
          "id": "0x1",
        },
        Object {
          "className": "TestSchema:TestClass",
          "id": "0x1",
        },
      ],
      "pathFromRoot": Array [
        "parentHash",
        "childHash",
      ],
      "type": "ECInstancesNode",
      "version": 2,
    },
    "label": PropertyRecord {
      "autoExpand": undefined,
      "description": undefined,
      "extendedData": undefined,
      "isDisabled": undefined,
      "isMerged": undefined,
      "isReadonly": undefined,
      "links": undefined,
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "test label",
        "value": "test_label",
        "valueFormat": 0,
      },
    },
    "parentId": "test_parent_id",
  },
]
`;
